File: apps/kimi-code/src/tui/components/dialogs/experiments-selector.ts
import {
  Container,
  Key,
  matchesKey,
  truncateToWidth,
  visibleWidth,
  type Focusable,
} from '@moonshot-ai/pi-tui';
import type { ExperimentalFeatureState } from '@moonshot-ai/kimi-code-sdk';

import { SELECT_POINTER } from '#/tui/constant/symbols';
import { currentTheme } from '#/tui/theme';
import { printableChar } from '#/tui/utils/printable-key';
import { SearchableList } from '#/tui/utils/searchable-list';

const ELLIPSIS = '…';

export interface ExperimentalFeatureDraftChange {
  readonly id: ExperimentalFeatureState['id'];
  readonly enabled: boolean;
}

export interface ExperimentsSelectorOptions {
  readonly features: readonly ExperimentalFeatureState[];
  readonly onApply: (changes: readonly ExperimentalFeatureDraftChange[]) => void;
  readonly onCancel: () => void;
}

export class ExperimentsSelectorComponent extends Container implements Focusable {
  focused = false;

  private readonly opts: ExperimentsSelectorOptions;
  private readonly list: SearchableList<ExperimentalFeatureState>;
  private readonly draft = new Map<ExperimentalFeatureState['id'], boolean>();

  constructor(opts: ExperimentsSelectorOptions) {
    super();
    this.opts = opts;
    this.list = new SearchableList({
      items: opts.features,
      toSearchText: (feature) => `${feature.title} ${feature.id} ${feature.description}`,
      searchable: true,
    });
  }

  handleInput(data: string): void {
    if (matchesKey(data, Key.escape)) {
      if (this.list.clearQuery()) return;
      this.opts.onCancel();
      return;
    }
    if (matchesKey(data, Key.enter)) {
      const changes = this.draftChanges();
      if (changes.length > 0) this.opts.onApply(changes);
      return;
    }
    const decoded = printableChar(data);
    if (matchesKey(data, Key.space) || decoded === ' ') {
      const selected = this.list.selected();
      if (selected !== undefined) this.toggleDraft(selected);
      return;
    }
    this.list.handleKey(data);
  }

  override render(width: number): string[] {
    const view = this.list.view();
    const titleSuffix =
      view.query.length === 0 ? currentTheme.fg('textMuted', '  (type to search)') : '';
    const hintParts = ['↑↓ navigate'];
    if (view.page.pageCount > 1) hintParts.push('PgUp/PgDn page');
    hintParts.push('Space toggle', 'Enter apply', 'Esc cancel');
    if (view.query.length > 0) hintParts.push('Backspace clear');

    const lines: string[] = [
      currentTheme.fg('primary', '─'.repeat(width)),
      currentTheme.boldFg('primary', ' Experimental features') + titleSuffix,
      currentTheme.fg('textMuted', ` ${hintParts.join(' · ')}`),
      '',
    ];

    if (view.query.length > 0) {
      lines.push(currentTheme.fg('primary', ` Search: `) + currentTheme.fg('text', view.query));
    }

    if (view.items.length === 0) {
      lines.push(currentTheme.fg('textMuted', '   No matches'));
    }

    for (let i = view.page.start; i < view.page.end; i++) {
      const feature = view.items[i]!;
      const selected = i === view.selectedIndex;
      lines.push(...this.renderFeature(feature, selected, width));
    }

    lines.push('');
    if (view.query.length > 0) {
      lines.push(
        currentTheme.fg(
          'textMuted',
          ` ${String(view.items.length)} / ${String(this.opts.features.length)}`,
        ),
      );
    } else if (view.page.end < view.items.length) {
      lines.push(
        currentTheme.fg(
          'textMuted',
          ` ▼ ${String(view.items.length - view.page.end)} more`,
        ),
      );
    }
    lines.push(this.renderApplyButton());
    lines.push(currentTheme.fg('primary', '─'.repeat(width)));
    return lines.map((line) => truncateToWidth(line, width, ELLIPSIS));
  }

  private toggleDraft(feature: ExperimentalFeatureState): void {
    if (isLocked(feature)) return;

    const enabled = !this.effectiveEnabled(feature);
    if (enabled === feature.enabled) {
      this.draft.delete(feature.id);
      return;
    }
    this.draft.set(feature.id, enabled);
  }

  private effectiveEnabled(feature: ExperimentalFeatureState): boolean {
    return this.draft.get(feature.id) ?? feature.enabled;
  }

  private isDraftChanged(feature: ExperimentalFeatureState): boolean {
    return this.effectiveEnabled(feature) !== feature.enabled;
  }

  private draftChanges(): ExperimentalFeatureDraftChange[] {
    const changes: ExperimentalFeatureDraftChange[] = [];
    for (const feature of this.opts.features) {
      if (this.isDraftChanged(feature)) {
        changes.push({ id: feature.id, enabled: this.effectiveEnabled(feature) });
      }
    }
    return changes;
  }

  private renderApplyButton(): string {
    const changes = this.draftChanges();
    const count = changes.length;
    const label = '[ Apply changes and reload ]';
    const summary =
      count === 0 ? 'no changes' : `${String(count)} ${count === 1 ? 'change' : 'changes'}`;
    const button = count === 0
      ? currentTheme.fg('textDim', label)
      : currentTheme.boldFg('primary', label);
    const summaryText = count === 0
      ? currentTheme.fg('textMuted', summary)
      : currentTheme.fg('success', summary);
    return ` ${button}  ${summaryText}`;
  }

  private renderFeature(
    feature: ExperimentalFeatureState,
    selected: boolean,
    width: number,
  ): string[] {
    const pointer = selected ? SELECT_POINTER : ' ';
    const prefix = currentTheme.fg(selected ? 'primary' : 'textDim', `  ${pointer} `);
    const label = selected ? currentTheme.boldFg('primary', feature.title) : currentTheme.fg('text', feature.title);
    const enabled = this.effectiveEnabled(feature);
    const status = enabled ? 'enabled' : 'disabled';
    const statusText = enabled ? currentTheme.fg('success', status) : currentTheme.fg('textDim', status);
    const detail = this.isDraftChanged(feature)
      ? `${featureDetail(feature)} · modified`
      : featureDetail(feature);
    const lines = [
      `${prefix}${label}  ${statusText}`,
      currentTheme.fg('textMuted', `    ${detail}`),
    ];
    const descriptionWidth = Math.max(1, width - 4);
    for (const line of wrapText(feature.description, descriptionWidth)) {
      lines.push(currentTheme.fg('textMuted', `    ${line}`));
    }
    return lines;
  }
}

function isLocked(feature: ExperimentalFeatureState): boolean {
  return feature.source === 'env' || feature.source === 'master-env';
}

function featureDetail(feature: ExperimentalFeatureState): string {
  const source = sourceLabel(feature);
  if (feature.source === 'env' || feature.source === 'master-env') {
    return `id ${feature.id} · ${source}`;
  }
  return `id ${feature.id} · ${source} · ${feature.env}`;
}

function sourceLabel(feature: ExperimentalFeatureState): string {
  switch (feature.source) {
    case 'master-env':
      return 'locked by DIMI_CODE_EXPERIMENTAL_FLAG';
    case 'env':
      return `locked by ${feature.env}`;
    case 'config':
      return 'config';
    case 'default':
      return 'default';
  }
}

function wrapText(text: string, width: number): string[] {
  const maxWidth = Math.max(1, width);
  const words = text
    .trim()
    .split(/\s+/)
    .filter((word) => word.length > 0);
  const lines: string[] = [];
  let current = '';

  for (const word of words) {
    const candidate = current.length === 0 ? word : `${current} ${word}`;
    if (visibleWidth(candidate) <= maxWidth) {
      current = candidate;
      continue;
    }
    if (current.length > 0) lines.push(current);
    current = visibleWidth(word) <= maxWidth ? word : truncateToWidth(word, maxWidth, ELLIPSIS);
  }

  if (current.length > 0) lines.push(current);
  return lines;
}

File: crates/dimi-tui/src/dialogs/experiments_selector.rs
//! Experimental features toggle list — port of
//! `apps/dimi/src/tui/components/dialogs/experiments-selector.ts`
//! (`ExperimentsSelectorComponent`).
//!
//! A searchable list where `Space` toggles each row's draft state in place;
//! `Enter` applies the accumulated changes. Rows from `env` / `master-env`
//! sources are locked and cannot be toggled.

use std::collections::HashMap;

use crate::component::Component;
use crate::keys::{decode_printable_key, matches_key};
use crate::searchable_list::SearchableList;
use crate::theme::{ColorToken, current_theme};
use crate::wrap::truncate_to_width;

use super::{SELECT_POINTER, wrap_text};

const ELLIPSIS: &str = "…";

/// `ExperimentalFeatureState['source']`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FeatureSource {
    Env,
    MasterEnv,
    Config,
    Default,
}

/// One experimental feature row.
#[derive(Debug, Clone)]
pub struct ExperimentalFeatureState {
    pub id: String,
    pub title: String,
    pub description: String,
    pub env: String,
    pub source: FeatureSource,
    pub enabled: bool,
}

/// A draft change (`ExperimentalFeatureDraftChange`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DraftChange {
    pub id: String,
    pub enabled: bool,
}

/// Action surfaced via [`ExperimentsSelectorComponent::take_action`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExperimentsSelectorAction {
    Apply(Vec<DraftChange>),
    Cancel,
}

/// `ExperimentsSelectorOptions`.
#[derive(Debug, Clone)]
pub struct ExperimentsSelectorOptions {
    pub features: Vec<ExperimentalFeatureState>,
}

/// `ExperimentsSelectorComponent`.
pub struct ExperimentsSelectorComponent {
    opts: ExperimentsSelectorOptions,
    list: SearchableList<ExperimentalFeatureState>,
    draft: HashMap<String, bool>,
    action: Option<ExperimentsSelectorAction>,
}

impl ExperimentsSelectorComponent {
    pub fn new(opts: ExperimentsSelectorOptions) -> Self {
        let list = SearchableList::new(
            opts.features.clone(),
            |feature| format!("{} {} {}", feature.title, feature.id, feature.description),
            None,
            None,
            true,
        );
        ExperimentsSelectorComponent {
            opts,
            list,
            draft: HashMap::new(),
            action: None,
        }
    }

    /// Host polls after `handle_input`.
    pub fn take_action(&mut self) -> Option<ExperimentsSelectorAction> {
        self.action.take()
    }

    fn is_locked(feature: &ExperimentalFeatureState) -> bool {
        matches!(
            feature.source,
            FeatureSource::Env | FeatureSource::MasterEnv
        )
    }

    fn toggle_draft(&mut self, feature: &ExperimentalFeatureState) {
        if Self::is_locked(feature) {
            return;
        }
        let enabled = !self.effective_enabled(feature);
        if enabled == feature.enabled {
            self.draft.remove(&feature.id);
            return;
        }
        self.draft.insert(feature.id.clone(), enabled);
    }

    fn effective_enabled(&self, feature: &ExperimentalFeatureState) -> bool {
        self.draft
            .get(&feature.id)
            .copied()
            .unwrap_or(feature.enabled)
    }

    fn is_draft_changed(&self, feature: &ExperimentalFeatureState) -> bool {
        self.effective_enabled(feature) != feature.enabled
    }

    fn draft_changes(&self) -> Vec<DraftChange> {
        let mut changes = Vec::new();
        for feature in &self.opts.features {
            if self.is_draft_changed(feature) {
                changes.push(DraftChange {
                    id: feature.id.clone(),
                    enabled: self.effective_enabled(feature),
                });
            }
        }
        changes
    }

    fn feature_detail(feature: &ExperimentalFeatureState) -> String {
        feature.detail()
    }

    fn render_apply_button(&self) -> String {
        let theme = current_theme();
        let changes = self.draft_changes();
        let count = changes.len();
        let label = "[ Apply changes and reload ]";
        let summary = if count == 0 {
            "no changes".to_owned()
        } else if count == 1 {
            "1 change".to_owned()
        } else {
            format!("{count} changes")
        };
        let button = if count == 0 {
            theme.fg(ColorToken::TextDim, label)
        } else {
            theme.bold_fg(ColorToken::Primary, label)
        };
        let summary_text = if count == 0 {
            theme.fg(ColorToken::TextMuted, &summary)
        } else {
            theme.fg(ColorToken::Success, &summary)
        };
        format!(" {button}  {summary_text}")
    }

    fn render_feature(
        &self,
        feature: &ExperimentalFeatureState,
        selected: bool,
        width: usize,
    ) -> Vec<String> {
        let theme = current_theme();
        let pointer = if selected { SELECT_POINTER } else { " " };
        let prefix = theme.fg(
            if selected {
                ColorToken::Primary
            } else {
                ColorToken::TextDim
            },
            &format!("  {pointer} "),
        );
        let label = if selected {
            theme.bold_fg(ColorToken::Primary, &feature.title)
        } else {
            theme.fg(ColorToken::Text, &feature.title)
        };
        let enabled = self.effective_enabled(feature);
        let status = if enabled { "enabled" } else { "disabled" };
        let status_text = if enabled {
            theme.fg(ColorToken::Success, status)
        } else {
            theme.fg(ColorToken::TextDim, status)
        };
        let detail = if self.is_draft_changed(feature) {
            format!("{} · modified", Self::feature_detail(feature))
        } else {
            Self::feature_detail(feature)
        };
        let mut lines = vec![
            format!("{prefix}{label}  {status_text}"),
            theme.fg(ColorToken::TextMuted, &format!("    {detail}")),
        ];
        let description_width = width.saturating_sub(4).max(1);
        for line in wrap_text(&feature.description, description_width) {
            lines.push(theme.fg(ColorToken::TextMuted, &format!("    {line}")));
        }
        lines
    }
}

fn source_label_for(feature: &ExperimentalFeatureState) -> String {
    match feature.source {
        FeatureSource::MasterEnv => "locked by DIMI_CODE_EXPERIMENTAL_FLAG".to_owned(),
        FeatureSource::Env => format!("locked by {}", feature.env),
        FeatureSource::Config => "config".to_owned(),
        FeatureSource::Default => "default".to_owned(),
    }
}

impl ExperimentalFeatureState {
    /// `featureDetail` — `id <id> · <source>` (env/master-env) or
    /// `id <id> · <source> · <env>`.
    fn detail(&self) -> String {
        if matches!(self.source, FeatureSource::Env | FeatureSource::MasterEnv) {
            format!("id {} · {}", self.id, source_label_for(self))
        } else {
            format!("id {} · {} · {}", self.id, source_label_for(self), self.env)
        }
    }
}

impl Component for ExperimentsSelectorComponent {
    fn render(&mut self, width: usize) -> Vec<String> {
        let theme = current_theme();
        let view = self.list.view();
        let title_suffix = if view.query.is_empty() {
            theme.fg(ColorToken::TextMuted, "  (type to search)")
        } else {
            String::new()
        };
        let mut hint_parts = vec!["↑↓ navigate"];
        if view.page.page_count > 1 {
            hint_parts.push("PgUp/PgDn page");
        }
        hint_parts.push("Space toggle");
        hint_parts.push("Enter apply");
        hint_parts.push("Esc cancel");
        if !view.query.is_empty() {
            hint_parts.push("Backspace clear");
        }

        let mut lines: Vec<String> = vec![
            theme.fg(ColorToken::Primary, &"─".repeat(width)),
            format!(
                "{}{}",
                theme.bold_fg(ColorToken::Primary, " Experimental features"),
                title_suffix
            ),
            theme.fg(
                ColorToken::TextMuted,
                &format!(" {}", hint_parts.join(" · ")),
            ),
            String::new(),
        ];

        if !view.query.is_empty() {
            lines.push(format!(
                "{}{}",
                theme.fg(ColorToken::Primary, " Search: "),
                theme.fg(ColorToken::Text, &view.query)
            ));
        }

        if view.items.is_empty() {
            lines.push(theme.fg(ColorToken::TextMuted, "   No matches"));
        }

        for i in view.page.start..view.page.end {
            if let Some(feature) = view.items.get(i) {
                let selected = i == view.selected_index;
                lines.extend(self.render_feature(feature, selected, width));
            }
        }

        lines.push(String::new());
        if !view.query.is_empty() {
            lines.push(theme.fg(
                ColorToken::TextMuted,
                &format!(" {} / {}", view.items.len(), self.opts.features.len()),
            ));
        } else if view.page.end < view.items.len() {
            lines.push(theme.fg(
                ColorToken::TextMuted,
                &format!(" ▼ {} more", view.items.len() - view.page.end),
            ));
        }
        lines.push(self.render_apply_button());
        lines.push(theme.fg(ColorToken::Primary, &"─".repeat(width)));
        lines
            .iter()
            .map(|line| truncate_to_width(line, width, ELLIPSIS, false))
            .collect()
    }

    fn handle_input(&mut self, data: &str) {
        if matches_key(data, "escape") {
            if self.list.clear_query() {
                return;
            }
            self.action = Some(ExperimentsSelectorAction::Cancel);
            return;
        }
        if matches_key(data, "enter") {
            let changes = self.draft_changes();
            if !changes.is_empty() {
                self.action = Some(ExperimentsSelectorAction::Apply(changes));
            }
            return;
        }
        let decoded = decode_printable_key(data);
        if matches_key(data, "space") || decoded.as_deref() == Some(" ") {
            if let Some(selected) = self.list.selected() {
                self.toggle_draft(&selected);
            }
            return;
        }
        self.list.handle_key(data);
    }

    fn invalidate(&mut self) {}
}

#[cfg(test)]
mod tests {
    use super::*;

    fn feature(
        id: &str,
        title: &str,
        source: FeatureSource,
        enabled: bool,
    ) -> ExperimentalFeatureState {
        ExperimentalFeatureState {
            id: id.to_owned(),
            title: title.to_owned(),
            description: format!("{title} description"),
            env: format!("DIMI_EXP_{}", id.to_uppercase()),
            source,
            enabled,
        }
    }

    fn opts() -> ExperimentsSelectorOptions {
        ExperimentsSelectorOptions {
            features: vec![
                feature("steer", "Steer into turn", FeatureSource::Config, false),
                feature("swarm", "Swarm mode", FeatureSource::Default, true),
                feature("locked", "Locked flag", FeatureSource::Env, false),
            ],
        }
    }

    #[test]
    fn renders_rows_with_status() {
        let mut c = ExperimentsSelectorComponent::new(opts());
        let lines = c.render(80);
        let joined = crate::ansi::strip_ansi(&lines.join("\n"));
        assert!(joined.contains("Experimental features"), "{joined}");
        assert!(joined.contains("Space toggle"), "{joined}");
        assert!(joined.contains("Steer into turn"), "{joined}");
        assert!(joined.contains("enabled"), "{joined}");
        assert!(joined.contains("disabled"), "{joined}");
        assert!(
            joined.contains("id steer · config · DIMI_EXP_STEER"),
            "{joined}"
        );
        assert!(joined.contains("no changes"), "{joined}");
    }

    #[test]
    fn space_toggles_draft_and_apply_emits() {
        let mut c = ExperimentsSelectorComponent::new(opts());
        // select second feature (swarm, enabled) and toggle it off.
        c.handle_input("\x1b[B");
        c.handle_input(" ");
        c.handle_input("\r");
        match c.take_action() {
            Some(ExperimentsSelectorAction::Apply(changes)) => {
                assert_eq!(
                    changes,
                    vec![DraftChange {
                        id: "swarm".to_owned(),
                        enabled: false
                    }]
                );
            }
            other => panic!("expected apply, got {other:?}"),
        }
    }

    #[test]
    fn locked_features_ignore_space() {
        let mut c = ExperimentsSelectorComponent::new(opts());
        c.handle_input("\x1b[B");
        c.handle_input("\x1b[B"); // onto locked
        c.handle_input(" ");
        c.handle_input("\r");
        assert_eq!(c.take_action(), None); // no draft changes → no apply
    }

    #[test]
    fn escape_cancels() {
        let mut c = ExperimentsSelectorComponent::new(opts());
        c.handle_input("\x1b");
        assert_eq!(c.take_action(), Some(ExperimentsSelectorAction::Cancel));
    }

    #[test]
    fn detail_renders_sources() {
        let f = feature("x", "X", FeatureSource::Env, false);
        assert_eq!(f.detail(), "id x · locked by DIMI_EXP_X");
        let f2 = feature("y", "Y", FeatureSource::MasterEnv, false);
        assert_eq!(f2.detail(), "id y · locked by DIMI_CODE_EXPERIMENTAL_FLAG");
    }
}
